### PR TITLE
Describe volumes

### PR DESCRIPTION
**Pull request by robert-sandor**

### CODE REVIEW

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/entities/aws/Filter.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    public String getName() {
+        return name;
+    }
+
+    public List<String> getValues() {
+        return values;
+    }
+
+    @Override
```
Comment (mihaitusa) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/entities/aws/Filter.java]: @robert-sandor Just a curiosity... you used Apache?

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/entities/inputs/FilterInputs.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        public FilterInputs build() {
+            return new FilterInputs(this);
+        }
+
+        public Builder withDelimiter(String delimiter) {
+            this.delimiter = delimiter;
+            return this;
+        }
+    }
+}
```
Comment (mihaitusa) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/entities/inputs/FilterInputs.java]: @victorursan I like this one... recognize your style... ;0)

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/FilterUtils.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package io.cloudslang.content.amazon.factory.helpers;
```
Comment (mihaitusa) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/FilterUtils.java]: @robert-sandor  Nice!

File: cs-amazon/src/test/java/io/cloudslang/content/amazon/entities/aws/VolumeAttachmentStatusTest.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+/*******************************************************************************
```
Comment (mihaitusa) [inline cs-amazon/src/test/java/io/cloudslang/content/amazon/entities/aws/VolumeAttachmentStatusTest.java]: @robert-sandor Don't think is really necessary... 😄

**Comment by mihaitusa** [comment]
Glad to see that Amazon integration is growing... (using Java of-course) 😄

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/actions/volumes/DescribeVolumesAction.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+/**
+ * Created by sandorr
+ * 2/13/2017.
+ */
+public class DescribeVolumesAction {
+    /**
+     * Describes one or more volumes.
+     * Note : If you are describing a long list of volumes, you can paginate the output to make the list more manageable.
+     * The maxResults parameter sets the maximum number of results returned in a single page.
+     * If the list of results exceeds your MaxResults value, then that number of results is returned along with a
```
Comment (Tirla-Alin) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/actions/volumes/DescribeVolumesAction.java]: Use either maxResults or MaxResults.

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/actions/volumes/DescribeVolumesAction.java [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+     * @param filterAttachmentDevice              Optional - The device name that is exposed to the instance.
+     *                                            Example : "/dev/sda1"
+     * @param filterAttachmentInstanceId          Optional - The ID of the instance the volume is attached to.
+     * @param filterAttachmentStatus              Optional - The attachment state.
+     *                                            Valid Values : attaching | attached | detaching | detached
+     * @param filterAvailabilityZone              Optional - The Availability Zone in which the volume was created.
+     * @param filterCreateTime                    Optional - The time stamp when the volume was created.
+     * @param filterEncrypted                     Optional - The encryption status of the volume.
+     * @param filterSize                          Optional - The size of the volume, in GiB.
+     * @param filterSnapshotId                    Optional -
```
Comment (Tirla-Alin) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/actions/volumes/DescribeVolumesAction.java]: Reminder.

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/actions/volumes/DescribeVolumesAction.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+     *                                            Example: "2016-11-15"
+     *                                            Default: "2016-11-15"
+     * @param delimiter                           Optional - Delimiter that will be used.
+     * @param filterAttachmentAttachTime          Optional - The time stamp when the attachment initiated.
+     * @param filterAttachmentDeleteOnTermination Optional -  Whether the volume is deleted on instance termination.
+     * @param filterAttachmentDevice              Optional - The device name that is exposed to the instance.
+     *                                            Example : "/dev/sda1"
+     * @param filterAttachmentInstanceId          Optional - The ID of the instance the volume is attached to.
+     * @param filterAttachmentStatus              Optional - The attachment state.
+     *                                            Valid Values : attaching | attached | detaching | detached
```
Comment (Tirla-Alin) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/actions/volumes/DescribeVolumesAction.java]: I would use comma as a separator and "values" with lower "v".

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/actions/volumes/DescribeVolumesAction.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+import static io.cloudslang.content.amazon.entities.constants.Constants.AwsParams.HTTP_CLIENT_METHOD_GET;
+import static io.cloudslang.content.amazon.entities.constants.Constants.DefaultApiVersion.INSTANCES_DEFAULT_API_VERSION;
+import static io.cloudslang.content.amazon.entities.constants.Constants.DefaultApiVersion.VOLUMES_DEFAULT_API_VERSION;
+import static io.cloudslang.content.amazon.entities.constants.Constants.Ec2QueryApiActions.DESCRIBE_VOLUMES;
+import static io.cloudslang.content.amazon.entities.constants.Constants.Miscellaneous.EMPTY;
+import static io.cloudslang.content.amazon.entities.constants.Inputs.CommonInputs.*;
+import static io.cloudslang.content.amazon.entities.constants.Inputs.InstanceInputs.MAX_RESULTS;
+import static io.cloudslang.content.amazon.entities.constants.Inputs.InstanceInputs.NEXT_TOKEN;
+import static io.cloudslang.content.amazon.entities.constants.Inputs.VolumeInputs.*;
+import static io.cloudslang.content.amazon.entities.constants.Outputs.FAILURE;
```
Comment (Tirla-Alin) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/actions/volumes/DescribeVolumesAction.java]: Use cs-commons constants.

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/actions/volumes/DescribeVolumesAction.java [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+            return new QueryApiExecutor().execute(commonInputs, volumeInputs, filterInputs);
+        } catch (Exception e) {
+            return ExceptionProcessor.getExceptionResult(e);
+        }
+    }
+
+
+
+    public static void main(String[] args) {
```
Comment (Tirla-Alin) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/actions/volumes/DescribeVolumesAction.java]: Reminder.

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/entities/aws/VolumeAttachmentStatus.java [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+            return NOT_RELEVANT;
+        }
+
+        for (VolumeAttachmentStatus volumeAttachmentStatus : VolumeAttachmentStatus.values()) {
+            if (volumeAttachmentStatus.name().equalsIgnoreCase(input)) {
+                return volumeAttachmentStatus.name().toLowerCase();
+            }
+        }
+
+        throw new RuntimeException("Invalid attachment.status value: [" + input + "]. Valid values: attaching, attached, " +
```
Comment (Tirla-Alin) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/entities/aws/VolumeAttachmentStatus.java]: You know what @victorursan would say.

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/entities/aws/VolumeStatus.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+            return NOT_RELEVANT;
+        }
+
+        for (VolumeStatus volumeStatus : VolumeStatus.values()) {
+            if (volumeStatus.getKey().equalsIgnoreCase(input)) {
+                return volumeStatus.getKey().toLowerCase();
+            }
+        }
+
+        throw new RuntimeException("Invalid status value: [" + input + "]. Valid values: creating, available, in-use, deleting, deleted, error.");
```
Comment (Tirla-Alin) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/entities/aws/VolumeStatus.java]: Same as above.

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/FilterUtils.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        final String[] splitString = StringUtils.split(string, delimiter);
+        return Arrays.asList(splitString);
+    }
+
+    public static String getFilterNameKey(int index) {
+        return String.format("Filter.%d.Name", index + 1);
+    }
+
+    public static String getFilterValueKey(int index, int counter) {
+        return String.format("Filter.%d.Value.%d", index + 1, counter + 1);
```
Comment (Tirla-Alin) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/FilterUtils.java]: I would use FILTER_VALUE_FORMAT and FILTER_NAME_FORMAT.

File: cs-amazon/src/test/java/io/cloudslang/content/amazon/entities/aws/VolumeAttachmentStatusTest.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+ * Created by sandorr
+ * 2/14/2017.
+ */
+public class VolumeAttachmentStatusTest {
+
+    private static final String ATTACHING = "attaching";
+    private static final String ATTACHED = "attached";
+    private static final String DETACHING = "detaching";
+    private static final String DETACHED = "detached";
+    private static final String RANDOM_STRING = "randomString";
```
Comment (Tirla-Alin) [inline cs-amazon/src/test/java/io/cloudslang/content/amazon/entities/aws/VolumeAttachmentStatusTest.java]: WRONG_VALUE would be more meaningful for me.

File: cs-amazon/src/test/java/io/cloudslang/content/amazon/execute/QueryApiExecutorTest.java [hunk 14]
```diff
@@ -833,6 +851,12 @@ private LoadBalancerInputs getLoadBalancerInputs() {
                 queryParamsMap.put("SnapshotId", "snap-id");
                 queryParamsMap.put("AvailabilityZone", "us-east-1d");
                 break;
+            case "DescribeVolumes":
+                queryParamsMap.put("Filter.1.Name", "status");
+                queryParamsMap.put("Filter.1.Value.1", "in-use");
```
Comment (Tirla-Alin) [inline cs-amazon/src/test/java/io/cloudslang/content/amazon/execute/QueryApiExecutorTest.java]: I would suggest to add a filter with 2 values to test the splitting functionality as well.

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/FilterUtils.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+        }
+        return filtersQueryMap;
+    }
+
+    public static Map<String, String> getFilterQueryMap(final Filter filter, int index, final FilterValidator validator) {
+        final Map<String, String> filterQueryMap = new HashMap<>();
+        filterQueryMap.put(getFilterNameKey(index), filter.getName());
+        for (final String value : filter.getValues()) {
+            final String key = getFilterValueKey(index, filter.getValues().indexOf(value));
+            if (Objects.equals(filter.getName(), "tag")) {
```
Comment (victorursan) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/FilterUtils.java]: why is there a magic string ?

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/FilterUtils.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+                System.out.println(filter.getValues());
```
Comment (victorursan) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/FilterUtils.java]: System.out.println ? :laughing: :rofl:

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/VolumeUtils.java [hunk 17]
```diff
@@ -78,6 +76,40 @@
...
+    public Map<String, String> getDescribeVolumesQueryParamsMap(InputsWrapper wrapper) {
+        Map<String, String> queryParamsMap = new HashMap<>();
+        setCommonQueryParamsMap(queryParamsMap, wrapper.getCommonInputs().getAction(), wrapper.getCommonInputs().getVersion());
+
+        setOptionalMapEntry(queryParamsMap, MAX_RESULTS, wrapper.getVolumeInputs().getMaxResults(),
+                !NOT_RELEVANT.equalsIgnoreCase(wrapper.getVolumeInputs().getMaxResults()));
+        setOptionalMapEntry(queryParamsMap, NEXT_TOKEN, wrapper.getVolumeInputs().getNextToken(),
+                isNotBlank(wrapper.getVolumeInputs().getNextToken()));
+
+        VolumesFilterValidator volumesFilterValidator = new VolumesFilterValidator();
```
Comment (victorursan) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/VolumeUtils.java]: final ?

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/VolumeUtils.java [hunk 18]
```diff
@@ -78,6 +76,40 @@
         return queryParamsMap;
     }
 
+    public Map<String, String> getDescribeVolumesQueryParamsMap(InputsWrapper wrapper) {
+        Map<String, String> queryParamsMap = new HashMap<>();
```
Comment (victorursan) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/VolumeUtils.java]: final ?

File: cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/VolumeUtils.java [hunk 19]
```diff
@@ -78,6 +76,40 @@
...
+        return queryParamsMap;
+    }
+
+    private Map<String, String> getVolumedIdsQueryMap(InputsWrapper wrapper) {
+        final String[] volumeIds = getArrayWithoutDuplicateEntries(wrapper.getVolumeInputs().getVolumeIdsString(),
+                VOLUME_IDS_STRING, wrapper.getCommonInputs().getDelimiter());
+        final Map<String, String> volumeIdsQueryMap = new HashMap<>();
+        if (isNotEmpty(volumeIds)) {
+            final List<String> volumeIdsList = Arrays.asList(volumeIds);
+            for (String id : volumeIdsList) {
```
Comment (victorursan) [inline cs-amazon/src/main/java/io/cloudslang/content/amazon/factory/helpers/VolumeUtils.java]: final

File: cs-amazon/src/test/java/io/cloudslang/content/amazon/entities/aws/VolumeAttachmentStatusTest.java (hunk 4 above)
Comment (victorursan) [inline cs-amazon/src/test/java/io/cloudslang/content/amazon/entities/aws/VolumeAttachmentStatusTest.java]: for the coverage

**Comment by victorursan** [comment]
what's the coverage percentage ?

**Comment by robert-sandor** [comment]
75% on cs-amazon

**Comment by victorursan** [comment]
we could increase it ... but it's ok for the moment

**Pull request merged**